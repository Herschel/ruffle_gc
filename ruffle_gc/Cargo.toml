[package]
name = "ruffle_gc"
version = "0.1.0"
authors = ["Mike Welsh <mwelsh@gmail.com>"]
edition = "2021"

[dependencies]
bitflags = "1.2"
generational-arena = "0.2.8"
once_cell = "1.7"
ruffle_gc_derive = { path = "../ruffle_gc_derive" }

[dev-dependencies]
trybuild = "1.0"
