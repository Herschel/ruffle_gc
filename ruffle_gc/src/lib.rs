mod context;
mod gc;
mod lifetime;
mod root;
mod trace;
mod weak;

pub use context::GcContext;
pub use gc::{Gc, GcVtbl};
pub use lifetime::GcLifetime;
pub use root::{GcHeapRoot, GcRoot, GcRootData};
pub use trace::Trace;
pub use weak::GcWeak;

pub use ruffle_gc_derive::Gc;

pub(crate) use gc::{GcData, GcDataPtr, GcFlags};
pub(crate) use weak::WeakId;

/// Creates a new GC root on the stack.
#[macro_export]
macro_rules! pin_root {
    ($name:ident $(,)?) => {
        let mut $name = $name;
        let mut $name = unsafe { ruffle_gc::GcRoot::new($name) };
        #[allow(unused_mut)]
        let mut $name = $name.pin();
    };
}
