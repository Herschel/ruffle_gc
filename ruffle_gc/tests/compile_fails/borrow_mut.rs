#![allow(unused_variables)]

use ruffle_gc::{Gc, GcContext, GcHeapRoot};

fn main() {
    let mut ctx = GcContext::new().unwrap();

    let root1 = GcHeapRoot::new(Root::new(&mut ctx, "Testing1"));
    let root2 = GcHeapRoot::new(Root::new(&mut ctx, "Testing2"));

    // Okay: shared borrows of GcContext
    let s1 = root1.0.borrow(&ctx);
    let s2 = root2.0.borrow(&ctx);

    // Error: Attempt to mutably borrow context while above shared borrows still exist.
    let s2 = root2.0.borrow_mut(&mut ctx);

    println!("{}", s1);
}

#[derive(Gc)]
struct Root<'a>(Gc<'a, String>);

impl<'a> Root<'a> {
    fn new(ctx: &'a mut GcContext, s: impl Into<String>) -> Self {
        Self(ctx.allocate(s.into()))
    }
}
