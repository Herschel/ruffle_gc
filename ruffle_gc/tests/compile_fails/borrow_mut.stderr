error[E0502]: cannot borrow `ctx` as mutable because it is also borrowed as immutable
  --> tests/compile_fails/borrow_mut.rs:16:33
   |
12 |     let s1 = root1.0.borrow(&ctx);
   |                             ---- immutable borrow occurs here
...
16 |     let s2 = root2.0.borrow_mut(&mut ctx);
   |                                 ^^^^^^^^ mutable borrow occurs here
17 |
18 |     println!("{}", s1);
   |                    -- immutable borrow later used here
